include/pybind11/pytypes.h

```diff
@@ -923,6 +923,9 @@ class type : public object {
     /// See https://github.com/pybind/pybind11/issues/2486
     template<typename T>
     static type of() {return type(type::handle_of<T>(), borrowed_t{}); }
+
+    /// Return the type name
+    object name() const { return attr("__name__"); }
```

Comment:
I don't see the point of returning an `object`. Then you could just as well use `attr("__name__")`, no?

If you want to add something to the C++ API, why not make it return `std::string`?

Reply:
Yeah returning `std::string` is better, since for both types and functions their `__name__` must be string, otherwise Python will throw type errors...

Reply:
Well, I'm not entirely convinced of that either, actually, because it would be inconsistent with `cpp_function::name()`.

Reply:
I think we can remove that one later and replace it with a `function::type()` that also returns `std::string`. It will be an incompatible change though...

Reply:
No, we cannot. That's a breaking change and cannot be done before pybind11 3.0, without getting error reports.

Reply:
How about just leave it as-is for now, and change it once pybind11 reaches 3.0?

Reply:
Where it's used, if it's not a `const char *`, it's converted back to a Python object. So it's not as simple as that.